PL-2247: Make joy promote release PR body more useful

silphid opened the pull request:
Add useful information and links (compare, Argo CD, Datadog) to all release promotion PRs.
Requires other PR:
https://github.com/nestoca/catalog/pull/980

davidmdm commented on internal/release/promote/perform.go:
```diff
@@ -116,30 +144,34 @@ func getPromotionMessage(releaseName string, sourceRelease, targetRelease *v1alp
-		// Put details of single promotion on first and only line
-		return fmt.Sprintf("%s (%s -> %s)", messages[0], sourceEnv, targetEnv)
+func getCommitMessage(pullRequestTemplate string, info *PromotionInfo) (string, error) {
+	tmpl, err := template.New("commit").Parse(pullRequestTemplate)
```
What if there's no pullRequestTemplate? We need a default!

davidmdm replied:
It might also make sense to parse the template before entering the command, this way we don't do all this work to realize that the template isn't parsable!

j-martin commented on internal/github/github_author.go:
```diff
@@ -0,0 +1,20 @@
+		repository = fmt.Sprintf("%s/%s", defaultOrganization, project.Name)
+	}
+
+	out, err := executeAndGetOutput(".", "api", "--jq", ".author.login", "repos/"+repository+"/commits/"+commitSHA)
```
This means we'd need to make an API query per commit. Can we do a single query for the commit range instead?

https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits:~:text=expects%20the%20format-,BASE...HEAD,-.%20Both%20must%20be

silphid replied:
That's definitely possible, but would require a significant bit of code to be refactored. I would think of that as premature optimization and revisit later if it proves to be an issue.

j-martin replied:
I just tested the latest version. Fixing the only-one-commit bug did make the PR creation substantially slower (~15 seconds). So slow I was wondering if the process hanged.

```
JOY_DEV_SKIP_VERSION_CHECK=1 go run ./cmd/joy release promote --no-prompt  --source qa --target demo products --dry-run
```

I think it's good enough for now (let's merge), but I disagree with it being a premature optimization.

For now can we add a log message explaining what joy is doing?

```
Fetching commits details from GitHub...
```

j-martin commented on internal/release/promote/perform.go:
```diff
@@ -107,45 +192,159 @@ func (p *Promotion) perform(params PerformParams) (string, error) {
+	uniqueAuthors := make(map[string]string)
+	for _, release := range info.Releases {
+		for _, commit := range release.Commits {
+			uniqueAuthors[commit.GitHubAuthor] = commit.GitHubAuthor
```
Great! That's going to be very useful to be notified when the PR is ready.

j-martin commented on internal/release/promote/perform.go:
```diff
@@ -107,45 +192,159 @@ func (p *Promotion) perform(params PerformParams) (string, error) {
+	for _, reviewer := range uniqueAuthors {
+		reviewers = append(reviewers, reviewer)
+	}
+	sort.Strings(reviewers)
```
Not that it matters, but I son't think we need to sort it.

silphid replied:
The order would be random. I prefer to have consistent, predictable and idempotent results, which is essential for unit/integration tests, but also makes for a better UX that can be explained.

j-martin replied:
Fair enough, it's not like this will be overly expensive. Even more so compared to fetching the authors in the first place. ;)

j-martin commented on internal/release/promote/perform.go:
```diff
@@ -1,98 +1,183 @@
+	if opts.dryRun {
+		fmt.Printf("ℹ️ Dry-run: skipping creation of branch %s\nFiles:\n%s\nCommit message:\n%s\n",
+			style.Resource(branchName), style.SecondaryInfo("- "+strings.Join(promotedFiles, "\n- ")),
+			style.SecondaryInfo(commitMessage))
```
Perfect. Let's add this in a comment in the `joy.yaml` in https://github.com/nestoca/catalog/pull/980 on how to preview the template.

silphid replied:
Good point, I just added that comment:
```
#
# Templates for `joy release promote` Pull Requests and Commits
# 
# When customizing those templates, you can use a command similar to the following to quickly iterate and test rendering locally:
# $ joy release promote --dry-run --skip-catalog-update --draft --no-prompt --source qa --target demo release1
#
```

j-martin reviewed (commented):
I have tried the PR. It looks like only one of commits made it in. There should be quite a few. https://github.com/nestoca/products/compare/api/v0.146.0...api/v0.149.0

I have not dug deeper than that.

```
➜  joy git:(feat(PL-2247)/better-pr-body) ✗ JOY_DEV_SKIP_VERSION_CHECK=1 go run ./cmd/joy release promote  --source qa --target demo products --dry-run
ℹ️ Dry-run mode enabled: No changes will be made.
————————————————————————————————————————————————————————————————————————————————
🚀 Update release demo/products (/Users/jm/src/github.com/nestoca/catalog/environments/demo/releases/origination/products.yaml)
nesto.ca/notifications-channel: notif-cicd-origination
 spec:
   project: products
-  version: 0.146.0
+  version: 0.149.0
   chart:
     version: 1.5.0
   values:
————————————————————————————————————————————————————————————————————————————————
? Select state of promotion PR? Ready
? Do you want to auto-merge the resulting PR? No
🚀 Updating release demo/products (/Users/jm/src/github.com/nestoca/catalog/environments/demo/releases/origination/products.yaml)
ℹ️ Dry-run: skipping writing promoted release products to: /Users/jm/src/github.com/nestoca/catalog/environments/demo/releases/origination/products.yaml
ℹ️ Dry-run: skipping creation of branch promote-products-from-qa-to-demo-96740fbd-cf0e-4bf8-afdc-70c05bebe4f8
Files:
- /Users/jm/src/github.com/nestoca/catalog/environments/demo/releases/origination/products.yaml
Commit message:
Promote products 0.146.0 -> 0.149.0 (qa -> demo)
- Promote products 0.146.0 -> 0.149.0 https://github.com/nestoca/products/compare/api/v0.146.0...api/v0.149.0
ℹ️ Dry-run: skipping creation of pull request:
Promote products 0.146.0 -> 0.149.0 (qa -> demo)
# Promotions
| Release | Diff | Source: qa | Target: demo |
|:---|:---|:---|:---|
| `products` | [Upgrade](https://github.com/nestoca/products/compare/api/v0.146.0...api/v0.149.0) | 0.149.0 <br> [Argo CD](https://argocd.platform.nesto.ca/applications/argocd/qa-products) <br> [Datadog](https://app.datadoghq.com/logs?query=env%3Aqa%20service%3Aproducts&cols=env%2Cservice) | 0.146.0 <br> [Argo CD](https://argocd.platform.nesto.ca/applications/argocd/demo-products) <br> [Datadog](https://app.datadoghq.com/logs?query=env%3Ademo%20service%3Aproducts&cols=env%2Cservice) |

# Upgrade commits for `products`
| SHA | Author | Message |
|:---|:---|:---|
| [59e3340](https://github.com/nestoca/products/commit/59e3340c7508ec66d36af4955524aa84b84f3df1) | @john-kohout-nesto | Merge pull request #364 from nestoca/feat/SERV-4253-update-ledgers |
Reviewers:
- john-kohout-nesto
🍺 Promotion complete!
```

Other than, it's looks pretty good.

davidmdm commented on cmd/joy/release.go:
```diff
@@ -326,82 +330,32 @@ func NewGitCommands() *cobra.Command {
 				}
 
-				fmt.Println(color.InCyan("running:"), color.InBold(strings.Join(gitArgs, " ")), color.InCyan("in"), color.InBold(repoDir))
+				fmt.Println(color.InCyan("running:"), color.InBold("git "+strings.Join(gitArgs, " ")), color.InCyan("in"), color.InBold(repoDir))
```
"git" was already included in the args. this will print: git git, which I've heard is rude in England.

silphid replied:
I'd rather say that what's a bit rude is naming a variable `gitArgs` and... putting more than just args in it, just to make the print easier, to then later remove the extra stuff when you really just needs the args part! 😜  See, it even got myself confused as to what to expect to have in that variable...

davidmdm commented on internal/git/git.go:
```diff
@@ -238,6 +238,21 @@ func GetCurrentCommit(dir string) (string, error) {
+func Fetch(dir string) error {
+	var buffer bytes.Buffer
+
+	cmd := exec.Command("git", "fetch")
```
Technically you copied my code, but I just realized this is the same as doing `CombinedOutput` instead of `Run`.
Not a requested changed, because there are more than 1 road to rome, but yeah. Brain fart!

silphid replied:
Yeap, indeed! I just fixed both funcs, please double-check! 🙏

Also made me realize that with `cmd.Dir = dir` we can avoid all the `"-C", dir` everywhere in that file. Not that it matters that much, but just makes me wonder why I didn't go with that approach in the first place(!) 🤷

davidmdm commented on internal/release/promote/commits_metadata.go:
```diff
@@ -0,0 +1,46 @@
+	Message string
+}
+
+func GetCommitsMetadata(dir, from, to string) ([]*CommitMetadata, error) {
```
This is not a request for change, but these types of string parsing functions make me think that as soon as we get some bandwidth for maintenance week, we should introduce `go-git`

silphid replied:
Yeah, totally! 👍

silphid commented:
> I have tried the PR. It looks like only one of commits made it in. There should be quite a few. 
Good catch! There was indeed a bug in multi-commit releases, which is now fixed.

Note that I also just added a `--skip-catalog-update` flag (which is currently only allowed in dry-run mode, we can later discuss if we want to make it available in regular mode, but I felt it was too risky). It's super helpful during development, troubleshooting and for testing templates, otherwise I always had to commit and push all my catalog changes to the remote for every single little detail I wanted to try.

j-martin reviewed (approved):
Let's add a log message to explain what `joy` is doing and let's merge it.

```
Fetching commits details from GitHub...
```
